airbyte-commons/src/main/java/io/airbyte/commons/util/ParallelCompositeIterator.java

```diff
@@ -0,0 +1,276 @@
+ * parallel.
+ * <p>
+ * </p>
+ * This iterator submits each iterator in the set of iterators to a thread pool, which fetches data
+ * from each iterator in parallel up to a parallelism limit. Results are offered to an internal
+ * queue, which is drained by the main {@link #next()} method implemented by this iterator. Note
+ * that the {@link #next()} method call will block indefinitely waiting for data to become available
+ * on the internal queue, provided that the iterator itself has not been closed or finished.
+ * <p>
+ * </p>
+ * The iterator is considered to be finished if there are no underlying iterators to read from
+ * <b>OR</b> if all underlying iterators have finished reading data and have signaled that they have
+ * finished.
+ * <p>
+ * </p>
+ * This iterator will also report stream status as it processes the underlying iterators for each
+ * iterator that implements the {@link AirbyteStreamAware} interface.
+ *
+ * @param <T> The type of data returned by a call to {@link #next()} on the iterator and all
+ *        underlying iterators.
+ */
+public final class ParallelCompositeIterator<T> extends AbstractIterator<T> implements AutoCloseableIterator<T>, AirbyteStreamAware {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ParallelCompositeIterator.class);
+  private static final Integer MAX_CAPACITY = 1000;
+  private static final Integer MAX_THREADS = 10;
+  private static final Integer MIN_THREADS = 1;
+
+  private final Optional<Consumer<AirbyteStreamStatusHolder>> airbyteStreamStatusConsumer;
+  private final ExecutorService executorService;
+  private boolean hasClosed;
+  private final BlockingQueue<InternalMessage<T>> internalQueue;
+  private final Set<AutoCloseableIterator<T>> iterators;
+  private boolean started = false;
+  private int closedCount = 0;
+
+  ParallelCompositeIterator(final Collection<AutoCloseableIterator<T>> iterators,
+                            final Consumer<AirbyteStreamStatusHolder> airbyteStreamStatusConsumer) {
+    this.airbyteStreamStatusConsumer = Optional.ofNullable(airbyteStreamStatusConsumer);
+    this.iterators = iterators != null ? new HashSet<>(iterators) : new HashSet<>();
+    /*
+     * TODO What is the correct upper bound here? It should probably be provided by the code that
+     * creates this iterator to account for the different types of connectors and their limitations. For
+     * instance, a JDBC-based source connector should probably set the "MAX_THREADS" to the maximum
+     * number of connections in the database connection pool used by the source. For API-based source
+     * connectors, this should be the maximum number of concurrent connections from one host/IP address
+     * that the API supports.
+     */
+    executorService = Executors.newFixedThreadPool(computeThreadPoolSize());
+    /*
+     * TODO Expose the max capacity as an advanced connector setting. This would allow the end user to
```

Comment:
I agree with this, I wonder if it should be a configuration shared between the platform and the connector. My reasoning for sharing this config is that the platform orchestrator will also implement an internal buffer and will likely face the same issues than the connector if the records are too big. WDYT?

Reply:
@benmoriceau Good point.  I'm unclear at the moment as to whether or not an end user would even understand what this lever does.  At a minimum, it should probably be in sync with the platform to ensure that we don't overrun it.  That being said, the connector also has different memory restrictions than the platform, so the setting may not be equal to what the platform uses, but some smaller value or dynamic value that goes down on retry if the previous failure is due to an OOM.  In any event, it needs to be configurable regardless of who sets the configuration.